### ISSUE TITLE
Support for Unix Sockets

### ISSUE THREAD

**Issue by SecurityInsanity** [sagebind/isahc#150]
:wave: Is there currently any support for Unix Sockets, or is this something not yet exposed (since libcurl does have a method for setting unix sockets)?

I'm trying to use this crate through [surf](https://github.com/http-rs/surf), and am running into an [issue](https://github.com/http-rs/surf/issues/137) when trying to use a unix based url:

```rust
surf::get("unix://var/run/docker.sock:0/v1.30/version")
```

produces the following logs:

```text
Jan 08 00:30:28.069 DEBUG clean:docker_api_get{path="/version" timeout=None}: dev_loop::executors::docker: URL for get will be: unix://var/run/docker.sock:0/v1.30/version
Jan 08 00:30:28.069 DEBUG clean:docker_api_get{path="/version" timeout=None}: isahc::agent: agent waker listening on 127.0.0.1:48125    
Jan 08 00:30:28.070 DEBUG isahc::agent: agent took 924.736µs to start up    
Jan 08 00:30:28.070 DEBUG isahc::agent: agent ready    
Jan 08 00:30:28.070  INFO clean:docker_api_get{path="/version" timeout=None}: surf::middleware::logger::native: sending request    
Jan 08 00:30:28.070 DEBUG isahc::handler: initializing handler for request [id=0]    
Jan 08 00:30:28.070 DEBUG isahc::curl: Protocol "unix" not supported or disabled in libcurl    
Jan 08 00:30:28.070 DEBUG isahc::curl: Closing connection -1    
Jan 08 00:30:28.070 DEBUG isahc::handler: curl error: [1] Unsupported protocol    
Jan 08 00:30:28.070  WARN isahc::handler: request completed with error [id=AtomicCell { value: 0 }]: Curl("Unsupported protocol"): Unsupported protocol    
Jan 08 00:30:28.071 DEBUG isahc::agent: agent shutting down    
Jan 08 00:30:28.071  WARN clean: dev_loop::executors::docker: Failed to reach out to docker engine api: [Failed to send docker API Request due to: [Curl("Unsupported protocol")]]
```

The response was _this seems to be something isahc needs to handle_:

```
Hi, thanks for opening this. It seems this goes correctly through surf but is being rejected by libcurl / isahc. Could you perhaps open an issue there linking back to this issue?

It seems in libcurl there's an option to instantiate unix sockets, which probably needs to be matched on through the url scheme: https://docs.rs/curl/0.5.0/curl/easy/struct.Easy2.html#method.unix_socket. This might be something that needs to be integrated on the isahc side to work.
```

Although it may be something isahc already handles, and surf just needs to do something with. Looking for any guidance :smile:

**Comment by sagebind** [comment]
Hi! No, there is not currently any support for unix sockets in Isahc, one reason being that support for them is currently disabled upstream when compiling a static libcurl. Here's the upstream issue: https://github.com/alexcrichton/curl-rust/issues/297.

Once that is resolved, we can add support for them here. It's definitely something that we ought to allow.

**Comment by sagebind** [comment]
So while https://github.com/alexcrichton/curl-rust/issues/297 is gearing up to becoming available, I am ruminating on what would be the proper way to expose this feature in Isahc. In @SecurityInsanity's original post, a URL like this is suggested:

```
unix:/{filepath}:0/{request-uri}
```

Is this standard practice? I am wondering what other HTTP clients do that support UNIX sockets. I've also seen URLs like

```
http+unix://{filepath}/{request-uri}
```

This seems to be the approach used by [HTTPie](https://github.com/httpie/httpie-unixsocket) and [requests-unixsocket](https://github.com/msabramo/requests-unixsocket). Some have mentioned that popular HTTP clients support something like this but so far my searching has turned up empty-handed. I'd love to see some prior art here if there is any.

Most clients that _do_ support UNIX sockets make it a separate configuration point entirely, since it is part of the transport layer and not the protocol layer (though I suppose newer HTTP versions muddy that water a bit). This includes libcurl itself, though we are not obligated to follow its lead. I'm actually leaning toward this last option just because it makes a bit more sense semantically.

Generalized, this feature is really just a way of saying, "For this specific request, connect to this address". DNS and hostname resolving is all ignored, and instead a connection is initiated with the address you specify. After that the HTTP protocol continues as normal. For UNIX sockets, this would be like a `.connect_to("unix://var/run/docker.sock")` option, but this could just as easily be generalized to support something like `.connect_to("tcp://127.0.0.1:1234")` to connect to a specific TCP socket.

If we want to be really flexible, we could allow the user to implement a trait that runs their own custom code to open a socket. Then in theory you could use a UNIX socket, Windows pipe, or whatever. Though that might be overcomplicating it.

**Comment by SecurityInsanity** [comment]
Hey @sagebind ,

Thanks for the detailed post! I originally suggested: `unix://` because I was looking at porting some ruby code which used: `unix://`: https://github.com/puppetlabs/net_http_unix. However it looks like most HTTP Clients (as you mentioned HTTPie, requests-unixsocket, and others use `http+unix`. So if we were going to support urls I'd say: `http+unix://` is the way to go as it seems more standard.

That being said I think the most common configuration point is separating it. That's what I've done in my fork: https://github.com/SecurityInsanity/isahc/commit/7706eb5a9cd07fa769fa205a51c0d7fd6a3be38e , what upstream curl-rust has done, and what many other http clients have done (restclient-cpp for C++, golang's http client, etc.)

I don't particularly like treating it as a proxy, and the flexibility sounds nice, but not sure if it's also needed for this? Seems like it could be done separately.

For me I think the "nicer" interface is: `http+unix://`, but if we wanted to be like more tools configuring it seperately seems to be the way to go.

**Comment by sagebind** [comment]
> I don't particularly like treating it as a proxy, and the flexibility sounds nice, but not sure if it's also needed for this? Seems like it could be done separately.

Perhaps, but I'd hesitate to rush out an API specifically for UNIX sockets if there's a more wholistic API that they fit into, and we end up deprecating the existing API in favor of. I like to ponder deep on Isahc's API (maybe a little _too_ deep) as I think the API quality is one of Isahc's best features. 😀

> For me I think the "nicer" interface is: `http+unix://`

At first blush, I do agree that encoding the socket path in the URL does seem to be the most convenient, and has the benefit that it is automatically supported by all downstream libraries and applications like surf, though with the caveat that we're forcing a particular scheme for it on downstream crates. The biggest downside with this approach is that the UNIX socket file path takes place of the hostname, meaning you lose some control over what the `Host` header is set to on the outgoing request. I _think_ manually setting `Host` would override correctly, but I'd have to double-check.

**Comment by SecurityInsanity** [comment]
Absolutely fair 😄 , just giving my two cents but I am admittedly biased as well 😉 .

**Comment by sagebind** [comment]
Interesting, I just found yet another convention used by [h2o](https://h2o.examp1e.net/configure/proxy_directives.html):

> In addition to TCP/IP over IPv4 and IPv6, the proxy handler can also connect to an HTTP server listening to a Unix socket. Path to the unix socket should be surrounded by square brackets, and prefixed with unix: (e.g. `http://[unix:/path/to/socket]/path`).

Perhaps that format makes the most sense of all, as it puts the UNIX socket info in the same place that IP socket info usually goes, though I wonder if `http::Uri` would reject that one.

**Comment by sagebind** [comment]
As another variable to consider, here's what `http::Uri` thinks of some of these custom URIs:

- `http://unix:/path/to/socket/path` -> host is `unix` (by the way, how can you tell where the UNIX path ends and the HTTP path begins on this one?)
- `http://unix:%2Fpath%2Fto%2Fsocket/path` -> `InvalidUri` error
- `http://[unix:/path/to/socket]/path` -> `InvalidUri` error
- `http://[unix:%2Fpath%2Fto%2Fsocket]/path` -> host is `[unix:%2Fpath%2Fto%2Fsocket]`
- `http://[unix:/path/to/socket]/path` -> `InvalidUri` error
- `unix:/path/to/socket:0/path` -> `InvalidUri` error
- `unix://path/to/socket:0/path` -> host is `path`
- `http+unix://path/to/socket/path` -> host is `path`
- `http+unix://path%2Fto%2Fsocket/path` -> `InvalidUri` error

Any syntax rejected by `http::Uri` can't really be used as all URIs pass through it in Isahc. The fact that `/` nor `%2F` are technically legal in the authority-part of a URL makes this a bit of a minefield.

**Comment by sagebind** [comment]
This capability is now finally available in the [0.9.7 release](https://github.com/sagebind/isahc/releases/tag/0.9.7) using the new [`Dialer`](https://docs.rs/isahc/0.9.7/isahc/config/struct.Dialer.html) API! Thanks for your patience and feedback.
